src/core/column/categorical.h

```diff
@@ -0,0 +1,70 @@
+//------------------------------------------------------------------------------
+// Copyright 2021 H2O.ai
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#ifndef dt_COLUMN_CATEGORICAL_h
+#define dt_COLUMN_CATEGORICAL_h
+
+#include "column/virtual.h"
+
+
+namespace dt {
+
+template <typename T>
+class Categorical_ColumnImpl : public Virtual_ColumnImpl {
+  private:
+    Buffer codes_;
+    Column categories_;
```

Comment:
There is a question of how to encode the NA values.
With your current approach, you only support NAs when they appear within the categories column. This could be too limited when one wants to set up categories to be a specific array. Also, this would make it harder to interoperate with the Arrow format.

The arrow's approach here is to have a separate `validity` buffer that uses a bitmask to encode whether each value is valid or not.

Reply:
>This could be too limited when one wants to set up categories to be a specific array. 

What do you mean? Let's say there is data `["cat", "dog", "mouse"]` and an array of categories `["cat", "dog"]`. Do you propose to encode `"mouse"` with `N/A`? I guess there could be a flag to enforce it, but by default the code should throw an exception saying that there is an unknown category.

Reply:
What i mean is suppose the categories are `["cat", "dog", "mouse"]`. Now, new data comes in, and it's, e.g., an `"aardvark"`. We don't have that among the categories, so this value has to be encoded as `NA`. But NA is also not among the categories, so what do you do now?

A more robust approach would be to allow the column of `codes` to be nullable, by adding an extra validity buffer.

Reply:
>so what do you do now?

Well, there are several options: either throw an error by saying there is an unknown category, or adding `"aardvark"` as a new category, or adding `N/A` as a new category and encode `"aardvark"` as `N/A`. I agree that having a separate buffer will simplify interconnection with arrow. We can always add another buffer with the N/A bitmask, but if instead of a buffer we can just add one element in the category column — why we shouldn't?

Reply:
In principle, if we establish convention that `N/A` (unknown categories) are always encoded as `0` — then, to create an N/A bitmask it would be enough just to cast the buffer of codes into `bool`. To go back, from the bitmask to the correct buffer or codes, it would be enough just to apply the mask to each element in the codes array.

Reply:
btw, arrow actually allows an `N/A` category

```
>>> pa.DictionaryArray.from_arrays([0, 1, 2, 0, 1], [None, "a", "b"])
<pyarrow.lib.DictionaryArray object at 0x7fc37ee31828>

-- dictionary:
  [
    null,
    "a",
    "b"
  ]
-- indices:
  [
    0,
    1,
    2,
    0,
    1
  ]
```